Return symbolic resource counts for dynamic control flow in pass-by-pass specs (#9507)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>New features since last release</h3>

> * :func:`~.specs` will now output symbolic resource information when it encounters a loop that uses dynamic control-flow
>   that can't be resolved at compile time.

Reviewer: ```suggestion
* :func:`~.specs` will now output symbolic resource information when it encounters a loop that uses dynamic control-flow
  that can't be resolved at compile time.
  In such cases the returned :class:`~.resource.CircuitSpecs` will contain :class:`~.resource.SymbolicSpecsResources`.
```

Author: ```suggestion
* :func:`~.specs` will now output symbolic resource information when it encounters a loop that uses dynamic control-flow
  that can't be resolved at compile time.
  In such cases the returned :class:`~.resource.CircuitSpecs` will contain :class:`~.resource.SymbolicSpecsResources` instances instead of the usual :class:`~.resource.SpecsResources` instances.
```

  ```python
  @qp.qjit(autograph=True)
  @qp.qnode(qp.device("lightning.qubit", wires=1))
  def circuit(x):
      qp.Hadamard(0)
      qp.PauliX(0)
      for _ in range(x):
          qp.PauliX(0)
      return qp.expval(qp.PauliX(0))

  specs_result = qp.specs(circuit, level=0)(5)
  ```

  ```pycon
  >>> print(specs_result)
  Device: lightning.qubit
  Device wires: 1
  Shots: Shots(total=None)
  Level: Before MLIR Passes

>   <BLANKLINE>

Reviewer 2: if `<BLANKLINE>` is just a new line, we just add new line here (empty line).

Author: Are you sure? I'm pretty sure `doctest` needs there to be a `<BLANKLINE>`: https://docs.python.org/3/library/doctest.html#how-are-docstring-examples-recognized

Reviewer 2: ahh i see. NP then.

  Symbolic Variables: a
  Wire allocations: 1
  Total gates: a + 2
  Gate counts:
  - Hadamard: 1
  - PauliX: a + 1
  Measurements:
  - expval(PauliX): 1
  Depth: Not computed
  ```

>   These symbolic resources have variables which can substituted for concrete values to compute the expected resources for a circuit.

Reviewer: ```suggestion
  These symbolic resources include expressions with variables which can substituted for concrete values to compute the associated resources for a circuit, via the ``subs`` method.
```
